databricks/koalas/indexing.py

```diff
@@ -400,16 +399,42 @@ def _select_rows(self, rows_sel):
                 # If slice is None - select everything, so nothing to do
                 return None, None
             elif len(self._internal.index_columns) == 1:
+                sdf = self._kdf_or_kser._internal.sdf
+                index_column = self._kdf_or_kser.index.to_series()
+                index_data_type = index_column.spark_type
                 start = rows_sel.start
                 stop = rows_sel.stop
 
-                index_column = self._kdf_or_kser.index.to_series()
-                index_data_type = index_column.spark_type
+                # get natural order from '__natural_order__' from start to stop
+                # to keep natural order when type of rows are StringType
+                if isinstance(index_data_type, StringType):
+                    start = sdf.select(NATURAL_ORDER_COLUMN_NAME) \
```

Comment:
@itholic, I mean the changes like this:

```diff
@@ -408,24 +395,21 @@ class LocIndexer(_LocIndexerLike):
                 # get natural order from '__natural_order__' from start to stop
                 # to keep natural order when type of rows are StringType
                 if isinstance(index_data_type, StringType):
-                    start = sdf.select(NATURAL_ORDER_COLUMN_NAME) \
-                               .where(index_column._scol == start) \
-                               .first()
-                    stop = sdf.select(NATURAL_ORDER_COLUMN_NAME) \
-                              .where(index_column._scol == stop) \
-                              .first()
-                    order_column = sdf[NATURAL_ORDER_COLUMN_NAME]
-                    if start is not None:
+                    start_and_stop = sdf.select(index_column._scol, NATURAL_ORDER_COLUMN_NAME) \
+                               .where((index_column._scol == start) | (index_column._scol == stop)) \
+                               .collect()
+
+                    start = [row[1] for row in start_and_stop if row[0] == start]
+                    if len(start) > 0:
                         start = start[0]
-                    elif rows_sel.start is None:
-                        pass
-                    else:
-                        raise KeyError(rows_sel.start)
-                    if stop is not None:
+                    stop = [row[1] for row in start_and_stop if row[0] == stop]
+                    if len(stop) > 0:
                         stop = stop[0]
-                    elif rows_sel.stop is None:
-                        pass
-                    else:
+
+                    order_column = sdf[NATURAL_ORDER_COLUMN_NAME]
+                    if start is None and rows_sel.start is not None:
+                        raise KeyError(rows_sel.start)
+                    if stop is None and rows_sel.stop is not None:
                         raise KeyError(rows_sel.stop)
                     index_data_type = LongType()
                 else:
```

Reply:
Ah, also add a comment that we call `collect()` assuming the frame does not have the records that have same start and end indexes a lot (otherwise it will cause out-of-memory easily)

Reply:
ah, that makes sense.
i though that you told like let's don't use job like `first()` or `collect()` like.
thanks for the comment with kindness example!